File: 在线仓库/modelzoo-master/samples/common/infer/post_process/crowdcount_postprocess.cpp
/*
 * Copyright (c) ModelZoo. 2025-2025. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "crowdcount_postprocess.h"
#include "log.h"
#include "utils.h"
#include <opencv2/opencv.hpp>
#include <fstream>
#include <vector>
#include <string>
#include <sys/stat.h>
#include <algorithm>
#include <numeric>

namespace Infer {
namespace CrowdCountNS {

// 常量定义
constexpr int CHANNEL_NUM = 1;
constexpr int TARGET_HEIGHT = 56;
constexpr int TARGET_WIDTH = 56;
constexpr float NORMALIZE_SCALE = 255.0f;

std::string GetFileName(const std::string& path) {
    size_t sep = path.find_last_of("/\\");
    size_t start = (sep == std::string::npos) ? 0 : sep + 1;
    size_t end = path.find_last_of(".");
    return (end == std::string::npos || end <= start) ?
           path.substr(start) : path.substr(start, end - start);
}

int CalculateRawCount(const cv::Mat& rawDensityMap) {
    float sum = 0.0f;
    for (int i = 0; i < rawDensityMap.rows; ++i) {
        const float* rowPtr = rawDensityMap.ptr<float>(i);
        for (int j = 0; j < rawDensityMap.cols; ++j) {
            sum += rowPtr[j];
        }
    }
    return static_cast<int>(sum);
}

std::pair<cv::Mat, int> PostprocessDensityMap(const float* preds, int targetHeight, int targetWidth,
                             int originalHeight, int originalWidth, float threshold) {
    // 转换为单通道密度图 (H, W) - 模型原始输出尺寸
    cv::Mat rawDensityMap(targetHeight, targetWidth, CV_32F);
    memcpy(rawDensityMap.data, preds, targetHeight * targetWidth * sizeof(float));

    // 应用阈值过滤
    rawDensityMap.setTo(0.0f, rawDensityMap < threshold);

    // 计算人数（在原始尺寸密度图上，与Python一致）
    int count = CalculateRawCount(rawDensityMap);

    // 为可视化准备缩放后的密度图
    cv::Mat visDensityMap = rawDensityMap.clone();
    if (originalHeight > 0 && originalWidth > 0 &&
        (originalHeight != targetHeight || originalWidth != targetWidth)) {
        cv::resize(visDensityMap, visDensityMap, cv::Size(originalWidth, originalHeight),
                  0, 0, cv::INTER_LINEAR);
    }

    return {visDensityMap, count};  // 返回可视化用图和计数结果
}

cv::Mat VisualizeDensityMap(const cv::Mat& densityMap) {
    // 归一化到0-255
    cv::Mat densityNorm;
    double maxVal;
    cv::minMaxLoc(densityMap, nullptr, &maxVal);

    if (maxVal > 0.0f) {
        densityMap.convertTo(densityNorm, CV_8UC1, NORMALIZE_SCALE / maxVal);
    } else {
        densityNorm = cv::Mat::zeros(densityMap.size(), CV_8UC1);
    }

    // 应用颜色映射
    cv::Mat coloredMap;
    cv::applyColorMap(densityNorm, coloredMap, cv::COLORMAP_JET);
    return coloredMap;
}

bool CrowdCountPostprocess(std::vector<std::string>& fileList,
                            std::vector<TensorBuf>& outBufs,
                            std::vector<TensorDesc>& outDescs) {
    LOG(INFO) << "====== 开始人群计数后处理 ======";
    auto cfg = ReadCfgFile("../data/cfg.txt");
    float threshold = std::stof(cfg["threshold"]);
    std::string outputDir = cfg["save_result_dir"];
    std::string saveBin = cfg["save_result_bin"];
    std::string densityMapDir = cfg["save_density_maps_dir"];
    std::string txtDir = outputDir + "/txt";

    // 创建输出目录结构
    if (!outputDir.empty()) {
        if (!CreateDirectoryRecursive(outputDir)) {
            LOG(ERROR) << "创建输出目录失败: " << outputDir;
            return false;
        }
        if (!saveBin.empty() && !CreateDirectoryRecursive(saveBin)) {
            LOG(ERROR) << "创建原始输出bin目录失败: " << saveBin;
            return false;
        }
        if (!densityMapDir.empty() && !CreateDirectoryRecursive(densityMapDir)) {
            LOG(ERROR) << "创建密度图目录失败: " << densityMapDir;
            return false;
        }
        if (!CreateDirectoryRecursive(txtDir)) {
            LOG(ERROR) << "创建TXT计数目录失败: " << txtDir;
            return false;
        }
    }

    // 处理每个图像
    for (size_t imgIdx = 0; imgIdx < fileList.size(); ++imgIdx) {
        const std::string& imgPath = fileList[imgIdx];
        std::string fileName = GetFileName(imgPath);
        LOG(INFO) << "处理文件: " << fileName;

        // 保存原始输出bin
        if (!saveBin.empty()) {
            std::string binPath = saveBin + "/" + fileName + "_output_0.bin";
            std::ofstream binFile(binPath, std::ios::binary);
            if (binFile.is_open()) {
                binFile.write(reinterpret_cast<char*>(outBufs[imgIdx].data.get()),
                             outBufs[imgIdx].size);
                LOG(INFO) << "保存原始输出bin至: " << binPath;
            } else {
                LOG(WARNING) << "无法保存原始输出bin至: " << binPath;
            }
        }

        // 验证输出数据尺寸
        const float* preds = reinterpret_cast<const float*>(outBufs[imgIdx].data.get());
        size_t expectedSize = CHANNEL_NUM * TARGET_HEIGHT * TARGET_WIDTH * sizeof(float);
        if (outBufs[imgIdx].size != expectedSize) {
            LOG(ERROR) << "输出尺寸不匹配，预期: " << expectedSize
                      << "，实际: " << outBufs[imgIdx].size;
            continue;
        }

        // 获取原始图像尺寸（即使读取失败也继续，使用模型输出尺寸可视化）
        int originalHeight = TARGET_HEIGHT;
        int originalWidth = TARGET_WIDTH;
        cv::Mat originalImg = cv::imread(imgPath);
        if (!originalImg.empty()) {
            originalHeight = originalImg.rows;
            originalWidth = originalImg.cols;
        } else {
            LOG(WARNING) << "无法读取原始图像: " << imgPath << "，使用模型输出尺寸可视化";
        }

        // 执行后处理（获取可视化图和计数结果）- 修复结构化绑定语法警告
        std::pair<cv::Mat, int> postResult = PostprocessDensityMap(
            preds, TARGET_HEIGHT, TARGET_WIDTH,
            originalHeight, originalWidth,
            threshold
        );
        cv::Mat visDensityMap = postResult.first;
        int count = postResult.second;

        // 输出人数计数（来自原始尺寸密度图的统计）
        LOG(INFO) << "人数计数: " << count;

        // 保存计数结果到单独的TXT文件（与Python逻辑一致）
        std::string txtPath = txtDir + "/" + fileName + ".txt";
        std::ofstream txtFile(txtPath);
        if (txtFile.is_open()) {
            txtFile << count;
            txtFile.close();
            LOG(INFO) << "已保存计数结果至: " << txtPath;
        } else {
            LOG(WARNING) << "无法保存计数结果至: " << txtPath;
        }

        // 可视化并保存密度图（使用缩放后的图）
        if (!densityMapDir.empty()) {
            cv::Mat densityMapVis = VisualizeDensityMap(visDensityMap);
            std::string imgSavePath = densityMapDir + "/" + fileName + "_density.jpg";
            if (!cv::imwrite(imgSavePath, densityMapVis)) {
                LOG(WARNING) << "保存密度图失败: " << imgSavePath;
            } else {
                LOG(INFO) << "已保存密度图至: " << imgSavePath;
            }
        }
    }

    LOG(INFO) << "所有后处理结果已保存至: " << outputDir;
    LOG(INFO) << "计数TXT文件已保存至: " << txtDir;
    LOG(INFO) << "====== 人群计数后处理完成 ======";
    return true;
}

}  // namespace CrowdCountNS
}  // namespace Infer

File: 在线仓库/modelzoo-master/samples/built-in/count/CrowdCount/script/crowdcount_postprocess.py
# Copyright (c) ModelZoo. 2025-2025. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import numpy as np
import cv2
import argparse
from PIL import Image


def postprocess(preds, original_h=None, original_w=None, threshold=0.01):
    """后处理：从模型输出计算人数并生成可视化密度图"""
    # 处理模型输出（密度图）
    density_map = preds.squeeze()  # 去除多余维度
    density_map[density_map < threshold] = 0  # 过滤低置信度值

    # 计算人数（密度图求和）
    count = int(np.sum(density_map))

    # 如果提供了原始尺寸且与当前尺寸不符，则进行缩放
    if original_h and original_w and (density_map.shape[0] != original_h or density_map.shape[1] != original_w):
        density_map = cv2.resize(density_map, (original_w, original_h), interpolation=cv2.INTER_LINEAR)

    # 密度图可视化处理（归一化到0-255）
    if np.max(density_map) > 0:
        density_map_norm = (density_map / np.max(density_map) * 255).astype(np.uint8)
    else:
        density_map_norm = np.zeros_like(density_map, dtype=np.uint8)
    density_map_colored = cv2.applyColorMap(density_map_norm, cv2.COLORMAP_JET)

    return count, density_map_colored


def run_postprocess(args):
    """执行后处理并保存结果（计数txt和密度图）"""
    # 创建输出目录结构
    os.makedirs(args.output_dir, exist_ok=True)
    os.makedirs(os.path.join(args.output_dir, 'density_maps'), exist_ok=True)
    os.makedirs(os.path.join(args.output_dir, 'txt'), exist_ok=True)

    # 获取所有推理结果文件
    bin_files = [f for f in os.listdir(args.bin_dir) if f.endswith('_output_0.bin')]
    if not bin_files:
        print(f"错误: 在 {args.bin_dir} 中未找到任何输出bin文件")
        return

    # 处理每个推理结果
    total = len(bin_files)
    target_h, target_w = args.target_size

    for idx, bin_file in enumerate(bin_files):
        # 提取图像名称
        img_name = bin_file.split("_output_0.bin")[0]

        # 获取原始图像尺寸（可选）
        original_h, original_w = None, None
        if args.img_dir:  # 仅当提供图像目录时才尝试读取原始尺寸
            img_path = os.path.join(args.img_dir, f"{img_name}.jpg")
            if os.path.exists(img_path):
                try:
                    img = Image.open(img_path)
                    original_w, original_h = img.size
                except Exception as e:
                    print(f"警告: 无法读取图像 {img_path}: {e}，将使用原始输出尺寸")
            else:
                print(f"警告: 原始图像 {img_path} 不存在，将使用原始输出尺寸")

        # 加载推理结果
        bin_path = os.path.join(args.bin_dir, bin_file)
        try:
            preds = np.fromfile(bin_path, dtype=np.float32)
            preds = preds.reshape(target_h, target_w)
        except Exception as e:
            print(f"警告: 处理文件 {bin_file} 失败: {e}，跳过")
            continue

        # 后处理
        count, density_map = postprocess(
            preds,
            original_h=original_h,
            original_w=original_w,
            threshold=args.threshold
        )

        # 保存密度图
        density_map_path = os.path.join(args.output_dir, 'density_maps', f"{img_name}_density.jpg")
        try:
            cv2.imwrite(density_map_path, density_map)
        except Exception as e:
            print(f"警告: 保存密度图 {density_map_path} 失败: {e}，跳过")
            continue

        # 保存计数结果到单独的txt文件
        txt_path = os.path.join(args.output_dir, 'txt', f"{img_name}.txt")
        try:
            with open(txt_path, 'w') as f:
                f.write(str(count))
        except Exception as e:
            print(f"警告: 保存计数结果 {txt_path} 失败: {e}，跳过")
            continue

        # 打印进度
        if (idx + 1) % 10 == 0 or (idx + 1) == total:
            progress = (idx + 1) / total * 100
            print(f"处理进度: {idx + 1}/{total} ({progress:.1f}%) - 文件: {bin_file} - 人数: {count}")

    print(f"所有后处理结果已保存至: {args.output_dir}")
    print(f"计数txt文件已保存至: {os.path.join(args.output_dir, 'txt')}")


def main():
    parser = argparse.ArgumentParser(description='CrowdCount后处理脚本')
    parser.add_argument('--bin_dir', default='../out/result_pc/bin',
                        help='存放推理输出bin文件的目录')
    parser.add_argument('--img_dir', default='../../../../../datasets/CrowDensity/Crowdata/img',
                        help='原始图像目录（可选，不提供则不进行缩放）')
    parser.add_argument('--output_dir', default='../out/result_pc',
                        help='输出结果目录（包含密度图和txt计数）')
    parser.add_argument('--target_size', type=int, nargs=2, default=[56, 56],
                        help='模型输出尺寸，格式为 高 宽，224x224输入对应56x56输出')
    parser.add_argument('--threshold', type=float, default=0.06,
                        help='密度图阈值过滤，默认0.06')

    args = parser.parse_args()

    # 验证输入目录
    if not os.path.isdir(args.bin_dir):
        print(f"错误: 推理结果目录 {args.bin_dir} 不存在")
        return
    run_postprocess(args)


if __name__ == "__main__":
    main()
